### ISSUE TITLE
Docker swarm - High latency for reaching a container after the swarm has been idle for a while (inner swarm container-to-container communication through dns)

### ISSUE THREAD

**Issue by augusto-altman** [moby/moby#35054]
**Description**

This issue is related to docker on swarm mode. We are experiencing high latency on inner container-to-container communication after the swarm has been idle for a while. 

Our configuration so far is really simple (docker-compose.yml): one network, one volume, 6 services 1 replica each (with the intention to have more replicas in the future), all services are rest services, only 1 node (with the intention to have more nodes in the future). After the swarm has been idle for a while (e.g. no incoming request on any container on a dev server) we experience high latency while reaching the different services in our swarm, but only for the first request on each of the services (only the first request is slow after being idle).

**Steps to reproduce the issue: (example)**
Lets assume we have a swarm with two services: Service A, Service B. Let's also assume that all the services' incoming requests (mentioned in the steps below) come from other containers through the inner swarm container communication system (dns). Then:

1. Idle 30 mins --> no incoming requests for any of the services during 30 mins
2. Incoming request (non-cacheable request) to Service A
3. Incoming request (non-cacheable request) to Service A
4. Incoming request (non-cacheable request) to Service B
5. Incoming request (non-cacheable request) to Service B

**Describe the results you received:**

1. Idle 30 mins (no incoming requests for any of the services during 30 mins)
2. Service A responds after (aprox) 20 secs
3. Service A responds immediately (and all further hits to Service A respond immediately)
4. Service B responds after (aprox) 20 secs
5. Service B responds immediately (and all further hits to Service B respond immediately)

**Describe the results you expected:**

1. Idle 30 mins (no incoming requests for any of the services during 30 mins)
2. Service A responds immediately
3. Service A responds immediately (and all further hits to Service A respond immediately)
4. Service B responds immediately
5. Service B responds immediately (and all further hits to Service B respond immediately)

**Additional information you deem important (e.g. issue happens only occasionally):**
As you can see above the issue only happens on the first request after the swarm has been idle for a while.
On the other hand we entered the containers (execute bash) and noticed that:

* Ping to other containers IP responds immediately
* Ping to other containers name (dns) has the mentioned delay (high latency)
* The container name (dns) resolution happens immediately (so it is not latency at dns level)

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:23:31 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 39
 Running: 5
 Paused: 0
 Stopped: 34
Images: 86
Server Version: 17.06.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 189
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 83vks7m0l7lll6jq6wqtudtq6
 Is Manager: true
 ClusterID: r6zc1q86ifvuza9bo5tqtmetf
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.1.66
 Manager Addresses:
  192.168.1.66:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.10.0-35-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.53GiB
Name: tito
ID: XELG:WLEI:GWUZ:YBPA:2G2W:W3G5:ABQS:GPYH:WHWO:3OMN:O4CB:633H
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: rav3nteam
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Host OS: Ubuntu 16.04

**Comment by pboers1988** [comment]
We have been experiencing similar problems, it seems like the overlay network is not responding. It needs to be reconfigured towards the node where the container is running, after an x amount of time. We experience the same symptoms regarding DNS resolution as described in the "Additional information" section.

Also running 17.06-ce however os is Centos 7

**Comment by augusto-altman** [comment]
@pboers1988 have you find any workaround?

**Comment by rgfretes** [comment]
we have the same behaviour on our end. 

We were NOT able to find a way around this delay.

**Comment by pboers1988** [comment]
@augusto-altman no workaround available unless you start employing bodges like pings from within containers. However I guess If you do full application functionality monitoring where something polls every 10 seconds you could solve the problem. This does not seem like a neat solution to me though....

**Comment by pboers1988** [comment]
Added options to keep the TCP socket alive on the intra container communication, will keep you posted about the results.

**Comment by augusto-altman** [comment]
@pboers1988 did the TCP socket alive thingy worked?

**Comment by pboers1988** [comment]
@augusto-altman Yes it seems to work. We are running a python project and keeping a postgres socket alive like this:

`          conn = pool.getconn()
            fd = conn.fileno()
            sock = socket.fromfd(fd, socket.AF_INET, socket.SOCK_STREAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 1)
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 3)
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 15)
`

**Comment by mbenalaya** [comment]
We have the same behavior on Ubuntu 16.04 docker version 17.09.0-ce

**Comment by st3xupery** [comment]
@pboers1988 I am thinking of applying a similar approach to this problem. Before I do so, have you approached the problem any differently as of late or have gained any new insight?

**Comment by st3xupery** [comment]
So the delays that plagued me were intra-container, specifically between a couple node applications (auth + api) and the MySQL container. I implemented a similar `keepAlive` technique for node mysql [here](https://github.com/mysqljs/mysql/issues/1939). 

As I mention in the thread I linked to, when my app is left idle for ~15 minutes the initial SQL query requests I record take ~100 seconds.

**Comment by sugun999** [comment]
We also have same issue with java spring-boot containers running with docker swarm.

**Comment by abjinugu** [comment]
We have similar issue. Here are the details.
3 Manger Nodes, 3 Worker Nodes (Windows) running legacy .Net Framework apps.
I'm using Newrelic to monitor my service performance.  I could see that my app takes insane amount of time to load when I hit it once in every 6-8 hours. However it is very quick if use it frequently. Its quick for every one else.
![when continuously used](https://user-images.githubusercontent.com/20075550/38169369-7d396ff2-3536-11e8-9add-d6785d096216.JPG)
![when used once in every 8 hours](https://user-images.githubusercontent.com/20075550/38169370-7d57aa4e-3536-11e8-90d9-f782eaabe053.JPG)

**Comment by pboers1988** [comment]
We solved this problem by tuning TCP keepalive kernel settings on Linux. The standard keepalive is 7200 seconds, however the internal swarm loadbalancer, ipvsadm, purges its connections after they become idle after 900 seconds. If you set the keep alive to something less than 900 seconds the problem will be solved.

‘’’
# cat /proc/sys/net/ipv4/tcp_keepalive_time 600 # cat /proc/sys/net/ipv4/tcp_keepalive_intvl 30# cat /proc/sys/net/ipv4/tcp_keepalive_probes 9
‘’’
Cheers,

Peter

**Comment by bastien612** [comment]
Can you explain this solution with more details ? It's hard to find documentation about "ipvsadm" and how Swarm use of it.
By default the 'file' you mention does not exist. Do you create them on all the hosts, only on the manager or into the image/containers ?
Does someone else tried this solution and can confirm that it works ?

**Comment by gtmadev** [comment]
Also seeing this issue. I thought maybe the mesh routing was just down, and then, after waiting about 7-10 seconds, it finally responds. Subsequent requests are fast and things work. I am thinking about ditching the overlay network idea for now and just use node placement labels to spread the nodes where I want, and then use HAPROXY to do the load balancing - at least until this is fixed.

**Comment by Drealine** [comment]
Hi,
Seems to have the same problem. The first request is very slow on Debian 10 with Docker Swarm and portainer.

Anyone to fix it ?

**Comment by grobacz** [comment]
This seems related [37822](https://github.com/moby/moby/issues/37822)
